docs/design/endpoints.md

```diff
@@ -0,0 +1,292 @@
+is a pre-serialization operation, and the rules language specifies several ways in which parameters are sourced.
+
+While provider parameters are defined per-service, the sources of the values may vary per-operation based on context
+parameters defined in the model.
+
+#### `clientContextParams`
+Service-level annotation. Defines an additional configuration field to be added to the service's client config. Bound to
+a field on the provider parameters at resolution time.
+
+#### `staticContextParams`
+Operation-level annotation. Defines a static value to be bound to a provider parameter at resolution time.
+
+#### `contextParams`
+Operation-level annotation. Marks a field on an operation to be bound to a provider parameter at resolution time.
+
+#### builtins
+The ruleset specification allows for the definition of builtins, or SDK-specific values that are bound to provider
+parameters at the time of resolution. As a generic codegen provider, smithy-kotlin doesn't implement any builtins on its
+own, but rather exposes an interface in the middleware generator for SDK authors to define and resolve their own
+builtins.
+
+For example, the AWS SDK has special builtins for the AWS region, as well as for concepts unique to the AWS ecosystem
+such as the usage of dual-stack endpoints. The AWS-specific generator implementation over top of the base smithy-kotlin
+one is responsible for providing the logic to code-generate the sourcing of these builtins in the resolver middleware
+(detailed below).
+
+## Codegen
+### Resolver
+Codegen for resolvers works much like any other extendable generated construct in smithy-kotlin - a base generator is
+implemented to perform to walk the rules tree and generate the core logic. This base class can be extended on by SDK
+authors as needed.
+
+```kotlin
+public open class EndpointProviderGenerator(
+    public val writer: KotlinWriter,
+    public val ruleset: EndpointRuleset,
+    // ...
+) {
+    /**
+     * Main codegen routine - walk the rules tree and generate the provider.
+     */
+    public fun render()
+
+    /**
+     * Resolve a standard library function referenced in a ruleset.
+     */
+    public open fun resolveFunction(func: String)
+}
+```
+
+### Middleware
+Endpoint resolution, whether generated or written by the caller, is performed per-request as a pre-operation middleware.
+At runtime, endpoint resolution is comprised of two principal tasks:
+1. construct a set of provider parameters, sourcing the values as specified by the model (context params)
+2. using the sourced parameters, call the provider to resolve to an endpoint, pointing the request to it
+
+Since endpoint resolution is performed on a per-operation basis, and the sourcing of provider parameters can vary
+per-operation, middleware must now be generated per-operation (this differs from the original implementation, where the
```

Comment:
Question: If an operation only uses service level context could there be a shared middleware for those operations?

This is a bit of premature optimization but we have seen how every additional class/lambda/etc adds up to overall artifact size. I would hazard a guess that most operations are not going to do anything fancy (since most don't now) and that generating a singular implementation that is shared across them would save a bunch on space/compilation/etc for larger models like DDB, EC2, S3, etc. It also would be backwards compatible because if a model ever added per/operation endpoint traits it would just generate a new internal middleware.